app/ui/browser-modern/sagas/session-sagas.js

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2016 Mozilla
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+this file except in compliance with the License. You may obtain a copy of the
+License at http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+import { takeLatest, throttle } from 'redux-saga';
+import { race, call, take, put, select } from 'redux-saga/effects';
+
+import { ipcRenderer } from '../../../shared/electron';
+import { wrapped } from './helpers';
+import { getWindowId } from '../selectors/ui';
+import { serializeAppState, deserializeAppState } from '../util/session-util';
+import * as ActionTypes from '../constants/action-types';
+import * as EffectTypes from '../constants/effect-types';
+import * as RootActions from '../actions/root-actions';
+import * as PageEffects from '../actions/page-effects';
+import * as SessionEffects from '../actions/session-effects';
+
+const THROTTLE_DURATION = 1000; // ms
+
+export default function*() {
+  yield [
+    takeLatest(...wrapped(EffectTypes.GET_SESSION_KEY, getSessionKey)),
+    takeLatest(...wrapped(EffectTypes.SET_SESSION_KEY, setSessionKey)),
+    takeLatest(...wrapped(EffectTypes.DEL_SESSION_KEY, delSessionKey)),
+    takeLatest(...wrapped(EffectTypes.RESTORE_SERIALIZED_APP_STATE, restoreSerializedAppState)),
+  ];
+
+  // Make sure we're cancelling the `watchAppStateChanges` task when either
+  // closing or reloading browser windows, so that everything is aborted even
+  // if recent actions cause the `saveAppStateToSession` task to start running
+  // due to throttling.
+  const winner = yield race({
+    task: call(watchAppStateChanges, saveAppStateToSession),
+    reload: take(EffectTypes.RELOAD_WINDOW),
+    close: take(EffectTypes.CLOSE_WINDOW),
+    stop: take(EffectTypes.STOP_SAVING_BROWSER_WINDOW_APP_STATE),
+  });
+
+  if (winner.close) {
+    // When closing a browser window, remove it from the session.
+    yield deleteAppStateFromSession();
+  } else if (winner.reload) {
+    // When reloading a browser window, save it one last time to the session,
+    // so that everything is guaranteed to be up to date. Otherwise, if the
+    // state changes just before throttling finishes, the very latest changes
+    // won't persist.
+    yield saveAppStateToSession();
+  }
+}
+
+function* getSessionKey({ key }) {
+  ipcRenderer.send('session:getkey', key);
```

Comment:
True, this isn't used anywhere. I'll remove it, though the main process does indeed handle the `session:getkey` event, not sure what you mean by "it doesn't return it".

The `session:setkey` and `session:delkey` events are indeed handled.